XEP-0055 - Add User Search

mightymop opened the pull request:
[XEP-0055: Jabber Search](https://xmpp.org/extensions/xep-0055.html)

sualko commented on scss/partials/_dialog.scss:
```diff
@@ -216,3 +216,22 @@
+  width: 400px;
+}
+
+.jsxc-searchresults-table thead tr th {
```
In Sass you can nest different rules to write less code.

```
.jsxc-searchresults-table {
  // your rules

  th { // th should be enough as selector
    // your th rules
  }
```

mightymop replied:
ok th and td now in the brackets of .jsxc-searchresults-table

sualko commented on src/connection/Form.ts:
```diff
@@ -91,14 +91,34 @@ export default class Form {
+                 this.reportedFields.forEach((field, index) => {
+                    let i=0;
+                    let found=false;
+                    for (;i<fields.length;i++)
```
i

sualko commented on src/connection/Form.ts:
```diff
@@ -91,14 +91,34 @@ export default class Form {
          if (fields.length !== this.reportedFields.length) {
-            throw new InvalidParameterError(`Item ${itemIndex} does not contain all "reported" fields.`);
+             try{
+                 this.reportedFields.forEach((field, index) => {
```
In which case is the server returning more fields than reported fields?

mightymop replied:
you are right ... maybe it was to late when i wrote the code...

sualko commented on src/connection/Form.ts:
```diff
@@ -91,14 +91,34 @@ export default class Form {
+                    }
+
+                    if (!found) {
+                       throw new InvalidParameterError(`Item ${itemIndex} does not contain all "reported" fields.`);
```
Why do we need the catch block, if we throw an error at this point?

mightymop replied:
ok it was to late :D

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,253 @@
+    dialog = new Dialog(content);
+    dialog.open();
+
+    let accountId = <string> $('#jsxc-account').val();
```
I guess this is the same as `account.getUid()`.

mightymop replied:
let accountId = <string> $('#jsxc-account').val();
let account = Client.getAccountManager().getAccount(accountId);

but need accountId to get account

sualko replied:
I guess it's the first account by default. Am I right?

mightymop replied:
we could also use: 
let account = Client.getAccountManager().getAccount()[0];

mightymop replied:
changed...

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,253 @@
+
+}
+
+function getSearchServiceDomain(connection : IConnection, account : Account)
```
Should be converted to an async function.

mightymop replied:
done

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,253 @@
+            {
+                //NO RESULTS FOUND!
+                dialog.close();
+                alert(Translation.t('no_results'));
```
`alert` should not be used.

mightymop replied:
ok now the dialog shows the error

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,253 @@
+            }
+            fields = $(result).find('query').find('reported').children('field');
+            let i=0;
+            for (;i<fields.length;i++)
```
i

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,253 @@
+
+    html+='</tbody></table>';
+
+    dialog.getDom().find('#content').empty().append(html);
```
Needs a prefix and should not use an id.

mightymop replied:
done

sualko reviewed (changes requested):
Nice addition, but the code is hard to read. A separation between simple and extended results should help.

mightymop commented:
> 
> 
> Nice addition, but the code is hard to read. A separation between simple and extended results should help.

tried to seperate some code now

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,278 @@
+export default function(username?: string) {
+
+    let content = contactSearchTemplate({
+      accounts: Client.getAccountManager().getAccounts().map(account => ({
```
The template is not using any variable.

mightymop replied:
Just a C&P error... will remove it

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,278 @@
+        html+='<td>';
+        if ($(items[i]).children('first').length>0)
+        {
+            html+=$(items[i]).children('first').text();
```
This allows XSS. We should not build any html from text.

mightymop replied:
changed to use jquery objects and methods now

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,278 @@
+        resolve(submitPromise);
+      });
+
+      formElement.find('.jsxc-js-close').click((ev) => {
+         ev.preventDefault();
+
+         dialog.close();
+
+         resolve(null);
+      });
```
The dialog close is looking for `.jsxc-js-close` and adding the close handler. So this can be removed.

mightymop replied:
without it does not close the dialog... maybe because its an dynamic form..

sualko commented on src/ui/dialogs/contactsearch.ts:
```diff
@@ -0,0 +1,278 @@
+
+        let formval = Form.fromHTML(formElement.get(0));
+
+        let submitPromise = connection.getSearchService().executeSearchForm(jid,formval).then((stanza) => {
+            return stanza;
+        });
+        resolve(submitPromise);
```
`return connection.getSearchService().executeSearchForm(jid,formval)` is just the same.

mightymop replied:
changed...

mightymop commented on src/connection/Form.ts:
```diff
@@ -90,15 +90,11 @@ export default class Form {
-            throw new InvalidParameterError(`Item ${itemIndex} does not contain all "reported" fields.`);
-         }
-
          this.reportedFields.forEach((field, index) => {
```
Because you mentioned in an earlier version: "In which case is the server returning more fields than reported fields?"
And i think we could trust the servers implementation because we started with the extended form, too.

sualko replied:
We can't trust the server, because it's not always "our" server. If there is no issue with the check, what is the point in removing it?

sualko commented:
I refactored the code, so that it is more readable and fixed multiple issue regarding simple/extended search/results. Also the user is not directly added, but rather the add contact dialog is shown. This way the user is able to configure the name (alias).

Currently I have to possibility to test this code, so it would be great if you could debug this. We also need to add a account selection.

mightymop commented:
> 
> 
> I refactored the code, so that it is more readable and fixed multiple issue regarding simple/extended search/results. Also the user is not directly added, but rather the add contact dialog is shown. This way the user is able to configure the name (alias).
> 
> Currently I have to possibility to test this code, so it would be great if you could debug this. We also need to add a account selection.

ok with last commit i added an account selection and fixed some bugs 
I changed from checkboxes in resulttable to radio because of the new starting dialog. dialog chaining did not work.
If you change the account, the form will be reloaded, if new searchservices will have other forms...

sualko commented on template/contactsearch.hbs:
```diff
@@ -1,5 +1,10 @@
 <div class="jsxc-content">
     {{t "Loading"}}
 </div>
-<div class="jsxc-results"></div>
```
What was the reason to remove the result area? This way the user could change his search without closing, opening the dialog.

mightymop replied:
re readded this... i did not thinkt about that ;)
